src/dma/mod.rs

```diff
@@ -465,115 +448,128 @@ where
-                    self.stream.set_memory_address(new_buf_ptr as u32);
-                }
-                // Check if an overrun occurred, the buffer address won't be
-                // updated in that case
-                if self.stream.get_memory_address() != new_buf_ptr as u32 {
-                    self.stream.clear_transfer_complete_interrupt();
-                    return Err(DMAError::Overrun(new_buf));
+            Some(inactive) => {
+                // Double buffer mode
+                if !STREAM::get_transfer_complete_flag() {
+                    // DMA has not released a buffer
+                    return Err(DMAError::NotReady);
                 }
-
-                // "Subsequent reads and writes cannot be moved ahead of
-                // preceding reads"
-                compiler_fence(Ordering::Acquire);
-
-                let old_buf = self.buf.replace(new_buf);
-
-                // We always have a buffer, so unwrap can't fail
-                return Ok((old_buf.unwrap(), CurrentBuffer::FirstBuffer));
-            } else {
+                // Poison the peripheral's inactive memory address to get a memory
+                // error instead of potentially silent corruption if the DMA peripheral
+                // wins the race to the inactive buffer.
+                // NOTE(safety): Memory corruption will still occur in the "incactive"
+                // buffer if an DMA overrun occurs between reading the CT bit and poisoning
+                // the address.
                 unsafe {
-                    self.stream
-                        .set_memory_double_buffer_address(new_buf_ptr as u32);
-                }
-                // Check if an overrun occurred, the buffer address won't be
-                // updated in that case
-                if self.stream.get_memory_double_buffer_address()
-                    != new_buf_ptr as u32
-                {
-                    self.stream.clear_transfer_complete_interrupt();
-                    return Err(DMAError::Overrun(new_buf));
+                    self.stream.set_memory_address(inactive, 0xffff_ffffusize);
                 }
-
-                // "Subsequent reads and writes cannot be moved ahead of
-                // preceding reads"
-                compiler_fence(Ordering::Acquire);
-
-                let old_buf = self.double_buf.replace(new_buf);
-
-                // double buffering, unwrap can never fail
-                return Ok((old_buf.unwrap(), CurrentBuffer::DoubleBuffer));
+                (true, inactive)
             }
-        }
-        self.stream.disable();
-        self.stream.clear_transfer_complete_interrupt();
+        };
 
-        // "No re-ordering of reads and writes across this point is allowed"
+        // Protect the instruction sequence of preceding DMA disable/inactivity
+        // verification/poisoning and subsequent (old completed) buffer content
+        // access.
         compiler_fence(Ordering::SeqCst);
```

Comment:
As @thalesfragoso has noted, it seems like just [`core::sync::atomic::fence(Ordering::SeqCst)`](https://doc.rust-lang.org/core/sync/atomic/fn.fence.html) is appropriate here to reduce the need to invoke two separate fences. This is true below as well wherever we have these fences